### ISSUE TITLE
About  transformed tables result

### ISSUE THREAD

**Issue by zhangbin1997** [luka-group/Lattice#8]
hi, the input format of my code in Totto is the same as that in your paper . I also shuffle the row and the column of table，but the baseline（T5-base）result only drops 0.2 -1.0 in bleu and parent, even if I exchange the order of metadata and table. In your parper, the T5-base drops 4.5 bleu. I don't know what's wrong with it.  thanks.
here is one example:
before shuffle：
'<unk>page_title> List of Governors of South Carolina <unk>/page_title> <unk>section_title> Governors under the Constitution of 1868 <unk>/section_title> <unk>table> <unk>cell> 76 <unk>h> # <unk>/h> <unk>h> 75 <unk>/h> <unk>h> 74 <unk>/h> <unk>/cell> <unk>cell> Daniel Henry Chamberlain <unk>h> 76 <unk>/h> <unk>h> Governor <unk>/h> <unk>/cell> <unk>cell> December 1, 1874 <unk>h> 76 <unk>/h> <unk>h> Took Office <unk>/h> <unk>/cell> <unk>/table></s>'
after shuffle:
'<unk>section_title> Governors under the Constitution of 1868 <unk>/section_title> <unk>table> <unk>cell> Daniel Henry Chamberlain <unk>h> Governor <unk>/h> <unk>h> 76 <unk>/h> <unk>/cell> <unk>cell> December 1, 1874 <unk>h> Took Office <unk>/h> <unk>h> 76 <unk>/h> <unk>/cell> <unk>cell> 76 <unk>h> # <unk>/h> <unk>h> 74 <unk>/h> <unk>h> 75 <unk>/h> <unk>/cell> <unk>/table> <unk>page_title> List of Governors of South Carolina <unk>/page_title></s>'

**Comment by zhangbin1997** [comment]
As shown in the example, i also change the order of related headers. I think the way can make the Totto dataset more harder, but the drop in T5-base is still small.

**Comment by FeiWang96** [comment]
Hi,

> the input format of my code in Totto is the same as that in your paper

No, it's not. At least the special markers are different. Note that our input format is designed for our method, where we simplify some layout markers because our proposed graph-based attention covers that information. For the T5 baseline, you can follow [this paper](https://arxiv.org/abs/2005.10433)'s implementation.

>  I also shuffle the row and the column of table，but the baseline（T5-base）result only drops 0.2 -1.0 in bleu and parent

As described in our paper, in addition to row and column shuffling, we also perturb the table layout by table transposing.

> I don't know what's wrong with it.

Given the current information, it is hard to say whether it's only the problem of your table layout perturbation or also the problem of baseline implementation.
To better understand what happened, please provide the following information:
* The code base you used;
* Data preprocessing code;
* Table perturbation scripts;
* BLEU scores on original/perturbed dev set.

**Comment by zhangbin1997** [comment]
### The code base you used;
> I completely use the code in your Lattice repo , including but not limited to T5 model and preprocess scripts.
### Data preprocessing code;
`the same to your repo, including but not limited to  lexicographic order, the only difference is that i replace <header> with <h>，replace </header> with </h> . I think the special markers difference will not influence the overall effect of the model. The reason I do this is to mitigate the influence of truncation.`
### Table perturbation scripts;
I shuffle the row and column order, and shuffle the row header and column header (I think this way of header shuffle is the same to table transposing in your paper)
For more details, here is one example.
Raw data（Json format is more convenient to display）：
```
'page_title': '<page_title> List of Governors of South Carolina </page_title> ', 
'section_title': '<section_title> Governors under the Constitution of 1868 </section_title> ', 
'table': {
         '<cell_3_0> 76': {
                       '<col_header>': {'<header> # </header> ': '', '<header> 75 </header> ': '', '<header> 74 </header> ': ''}}, 
         '<cell_3_2> Daniel Henry Chamberlain': {
                        '<row_header>': {'<header> 76 </header> ': ''}, 
                        '<col_header>': {'<header> Governor </header> ': ''}}, 
         '<cell_3_3> December 1, 1874': {
                       '<row_header>': {'<header> 76 </header> ': ''}, 
                       '<col_header>': {'<header> Took Office </header> ': }}
}
```
data without shuffle（will be tokenized to input_ids）:
`'<page_title> List of Governors of South Carolina </page_title> <section_title> Governors under the Constitution of 1868 </section_title> <table> <cell> 76 <h> # </h> <h> 75 </h> <h> 74 </h> </cell> <cell> Daniel Henry Chamberlain <h> 76 </h> <h> Governor </h> </cell> <cell> December 1, 1874 <h> 76 </h> <h> Took Office </h> </cell> </table></s>'`

data with shuffle（will be tokenized to input_ids）:
`'<section_title> Governors under the Constitution of 1868 </section_title> <table> <cell> December 1, 1874 <h> Took Office </h> <h> 76 </h> </cell> <cell> 76 <h> 74 </h> <h> 75 </h> <h> # </h> </cell> <cell> Daniel Henry Chamberlain <h> Governor </h> <h> 76 </h> </cell> </table> <page_title> List of Governors of South Carolina </page_title></s>'`

`I think the way I use includes the row-wise shuffling, column-wise shuffling and table transpose in your paper except the different special marker(<h>, <\h>). By the way, i also shuffle the metadata and table, because if i only shuffle table-inner information, the result drop will be less obvious;`

If my shuffle method is different from yours, please post an example, including before and after shuffle.

### BLEU scores on original/perturbed dev set.
for T5-base, Overall BLEU drops from 48.0 to 47.0, PARENT drops from 58.04 to 57.69. Overlap BLEU drops from 55.7 to 54.7, PARENT drops from 62.21 to 61.91.  Non-Overlap BLEU drops from 40.5 to 39.5, PARENT drops from 54.0 to 53.62.

I think the drop is too small, I intuitively judge that it is not normal on T5, but I haven't located the specific reason yet.

**Comment by FeiWang96** [comment]
### Baseline

If you want to reproduce the T5 baseline using our code, please follow these steps:
* Replace `preprocess_data_main.py` and `preprocess_utils.py` in our `preprocess` folder with the corresponding [baseline preprocessing code](https://github.com/google-research/language/tree/master/language/totto/baseline_preprocessing).
* Replace the [model, trainer, and tokenizer](https://github.com/luka-group/Lattice/blob/10560799ded087dd60b9b71bb2c2b5a90f09b850/train.py#L41) in `train.py` with the original ones:
  ```python
  from transformers import T5ForConditionalGeneration, T5TokenizerFast, Seq2SeqTrainer
  ```
This will let you reproduce the T5-based method proposed in [this paper](https://arxiv.org/abs/2005.10433).

### Table Layout Perturbation

To create the hard evaluation set, make sure you perturb the tables with row shuffling, column shuffling, and table transposing.
The last one is **not** equal to swapping column and row headers. Assume we linearize the following table by row-wise traversal

<div align="center">

| A| B|
|---|---|
| a1| b1 |
| a2| b2 |
| a3| b3 |

</div>

Let's ignore the headers and focus on the data cells. No matter how you swap the rows and columns, each a-b pair with the same subscript will always be next to each other in the linearized sequence (e.g. `b2 a2 b1 a1 b3 a3`). Table transposing will break such relations, resulting in a sequence where all a and b are separated (e.g. `b2 b1 b3 a2 a1 a3`).

Note that if you also perturb the tables in the training set, it will become the `Data Augmentation` method introduced in our paper.

**Comment by zhangbin1997** [comment]
I have a question. Are you using the another preprocessing code(baseline preprocessing code in Totto repo) and transformers library(from transformers import T5ForConditionalGeneration) for T5 in your experimental process? 
I read the code in your repo carefully. I think the preprocessing code in your repo can be used for both T5 and Lattice. And the model code(from model.modeling_t5 import T5ForConditionalGeneration) in your repo also can be used for both T5 and Lattice. For T5, we just set type_ids to None. For Lattice, we just reserve the parameter value of type_ids. Is there any deviation in my understanding. Or is there any other difference between the model code in your repo((from model.modeling_t5 import T5ForConditionalGeneration) ) and transformers library((from transformers import T5ForConditionalGeneration)).

Another question: you post an example of a-b pair, but the processing code is arranged by cell, such as `<cell1> *** </cell1> <cell2> *** </cell2> <cell3> *** </cell3>`. For each cell, there is no related pair information. I really can't understand the meaning of Table Layout Perturbation you mentioned. Can you give me a real example in dev set. Like the example i posed before, only needs (the sequence will be tokenized to input_ids without any shuffle and transpose) and (the sequence will be tokenized to input_ids with shuffle and transpose) . Thanks.

**Comment by FeiWang96** [comment]
The input format is part of a method. Vanilla T5 with our input format is the `layout-agnostic input` method introduced in our paper. Setting `type_id`, `row_id`, and/or `col_id` to `None` can lead to unexpected model behavior (e.g. activate a part of Lattice). If your goal is to develop your own method, then you can do whatever you want. However, if your goal is to reproduce existing methods, we highly suggest you follow our instructions. 


Below is an example of table layout perturbation with transpose. For simplicity, this figure only presents the subsequence for data cells.

<div align='center'>

<img width="500" alt="image" src="https://user-images.githubusercontent.com/19998174/212416400-ffb1ea06-d64f-4c15-8274-1ceecc45b188.png">

</div>

**Comment by zhangbin1997** [comment]
Thanks!
I follow your suggestion to reproduce T5 baseline result like following.
Replace preprocess_data_main.py and preprocess_utils.py in our preprocess folder with the corresponding baseline preprocessing code.
Replace the model, trainer, and tokenizer in train.py with the original ones: from transformers import T5ForConditionalGeneration, T5TokenizerFast, Seq2SeqTrainer.
but I I get a very high score. For T5-small, overall BLEU 48.7, overall PARENT 58.47. For T5-base, overall BLEU 49.1, overall PARENT 58.21. What is the reason? I use a larger batch size and smaller steps. for example, T5-base uses 16 batch size, 80000steps.

**Comment by FeiWang96** [comment]
That's weird. Those results are even much better than the T5-3B performance (48.4 BLEU and 57.8 PARENT) reported by [Kale and Rastogi, 2020](https://arxiv.org/abs/2005.10433). Can you try to reproduce their results with their hyper-parameters? 

I forget to mention that the [preprocess_function](https://github.com/luka-group/Lattice/blob/10560799ded087dd60b9b71bb2c2b5a90f09b850/train.py#L347) in `train.py` also needs to be cleaned (i.e. deleting all lines about `type_ids`, `col_ids`, and `row_ids`). However, that won't influence the performance as those params are not used.

**Comment by zhangbin1997** [comment]
I have deleted relevant type_ids, col_ids, and row_ids during the experiment. How do you reproduce the T5 baseline performance in your paper?  The performance of T5 baseline in your paper is basically equal to the performance reported by Kale and Rastogi, 2020.  Are you doing the same as you suggested me to do?

**Comment by FeiWang96** [comment]
That is our way to reproduce the baseline. However, I'm not sure whether you are doing exactly the same thing. Your T5-base results are even higher than the current SOTA method, which also uses T5-base as the backbone. That makes no sense. Regarding reproducing the T5 baseline, you can email the authors of the baseline paper directly.

**Comment by zhangbin1997** [comment]
Thanks. I have another question. For reproducing T5-baseline performance, do you also use the constant lr 2e-4 , batch size 8 and 150000 steps?  Because in Kale and Rastogi, 2020, it uses the constant 0.001 and 10K steps.

**Comment by FeiWang96** [comment]
Their hyper-parameters.

**Comment by zhangbin1997** [comment]
Thanks. Can you provide specific batch size, learning rate and training steps for T5 baseline in your experimental process?
